uninstall

```diff
@@ -0,0 +1,226 @@
+
+$HOMEBREW_FILES << HOMEBREW_CELLAR
+
+unless options[:skip_cache_log]
+  $HOMEBREW_FILES += %W[
+    #{ENV["HOME"]}/Library/Caches/Homebrew
+    #{ENV["HOME"]}/Library/Logs/Homebrew
+    /Library/Caches/Homebrew
+    #{ENV["HOME"]}/.cache/Homebrew
+  ].map { |p| Pathname.new(p) }
+end
+
+if /darwin/i === RUBY_PLATFORM
+  $HOMEBREW_FILES += %W[
+    /Applications
+    #{ENV["HOME"]}/Applications
+  ].map do |p|
+    Pathname.new(p).subdirs.select do |app|
+      app.resolved_path.to_s.start_with? HOMEBREW_CELLAR.to_s
+    end
+  end.flatten
+end
+
+$HOMEBREW_FILES.select!(&:exist?)
+
+unless options[:quiet]
+  warn "This script #{options[:dry_run] ? "would" : "will"} remove:"
+  $HOMEBREW_FILES.each(&:pretty_print)
+end
+
+unless options[:force] || options[:dry_run]
+  STDERR.print "Confirm to uninstall Homebrew? [y/N] "
+  abort unless gets.rstrip =~ /y|yes/i
+end
+
+ohai "Unlink Homebrew Installation..." unless options[:quiet]
+paths = %W[Frameworks bin etc include lib opt sbin share var].
+  map { |p| HOMEBREW_PREFIX/p }.select(&:exist?).map(&:to_s)
+args = %W[-E] + paths + %W[-regex .*/info/([^.][^/]*\.info|dir)]
+if options[:dry_run]
+  args << "-print"
+else
+  args += %W[-exec /bin/bash -c]
+  args << "/usr/bin/install-info --delete --quiet {} \"$(dirname {})/dir\""
+  args << ";"
+end
+puts "Would uninfo:" if options[:dry_run]
+system "/usr/bin/find", *args
+args = paths + %W[-type l -lname */Cellar/*]
+if options[:dry_run]
+  args << "-print"
+else
+  args += %W[-exec unlink {} ;]
+end
+puts "Would unlink:" if options[:dry_run]
+system "/usr/bin/find", *args
+
+ohai "Remove Homebrew Installation..." unless options[:quiet]
+$HOMEBREW_FILES.each do |file|
+  if options[:dry_run]
+    puts "Would remove #{file}"
+  else
+    FileUtils.rm_rf(file)
+  end
+end
+
+ohai "Cleanup..." unless options[:quiet]
+paths = %W[Frameworks bin etc include lib opt sbin share var].map { |p| HOMEBREW_PREFIX/p }
+paths << HOMEBREW_REPOSITORY/"Library"
+paths.select!(&:exist?)
+paths.map!(&:to_s)
+args = paths + %W[-name .DS_Store]
+if options[:dry_run]
+  args << "-print"
+else
+  args << "-delete"
+end
+puts "Would remove:" if options[:dry_run]
+system "/usr/bin/find", *args
+args = paths + %W[-depth -type d -empty]
+if options[:dry_run]
+  args << "-print"
+else
+  args += %W[-exec rmdir {} ;]
+end
+puts "Would remove directories:" if options[:dry_run]
+system "/usr/bin/find", *args
+# remove HOMEBREW_REPOSITORY HOMEBREW_PREFIX if they're empty.
+unless options[:dry_run]
+  Kernel.system "rmdir #{HOMEBREW_REPOSITORY} &>/dev/null"
+  Kernel.system "rmdir #{HOMEBREW_PREFIX} &>/dev/null"
+end
+
+ohai "Uninstallation done!" unless options[:quiet]
+
+residual_files = []
+residual_files.concat(HOMEBREW_REPOSITORY.subdirs) if HOMEBREW_REPOSITORY.exist?
+residual_files.concat(HOMEBREW_PREFIX.subdirs) if HOMEBREW_PREFIX.exist?
+residual_files.uniq!
+
+unless residual_files.empty? || options[:quiet]
+  puts "Following residual files found:"
```

Comment:
Technically, most residual files are the files we aren't sure about their original. Such as `HOMEBREW_PREFIX/lib/python2.7` or `HOMEBREW_PREFIX/var/log`.